### PR TITLE
Support evaluation on test set

### PR DESCRIPTION
**Pull request by himkt**
Adding support for evaluation on the test set.

### CODE REVIEW

**Comment by himkt** [comment]
@jun-harashima Could you please take a look? :bow:

File: msnc/model.py [hunk 1]
```diff
@@ -75,35 +71,66 @@ def _init_linears(self, linear_params):
             linears.append(linear)
         return nn.ModuleList(linears)
 
-    def run_training(self, output_dir, training_set, development_set=None):
+    def run_training(
+        self,
+        output_dir,
+        training_set,
+        development_set=None,
+        test_set=None,
```
Comment (jun-harashima) [inline msnc/model.py]: It's not a good idea to give a test set to a training method because of the following three reasons:

- A. the test set is essentially unnecessary for the training.
- B. it's easy to see the accuracies on the test set and tune our model according to the results.
- C. we may mistake the test set for the training or development sets when we change the codes. 🐛 

I can understand what you want to do. So, my idea is as follows:

- a. use the original code and select the best model on the development set for the test. This solves the above three issues.
- b. define a child class (e.g., `ModelForExperiment`) that has this `run_training` method with the `test_set` argument and use it for experiments. This slightly solves the issues A and C.

Comment (himkt) [reply msnc/model.py]: It's OK for me and I understand your opinion.
I revised my PR so that `run_training` does not take `test_set` and
take `save_best_model` to save `best.model` for testing.
What do you think about this change? :eyes:

File: msnc/model.py [hunk 2]
```diff
@@ -96,14 +99,32 @@ def run_training(self, output_dir, training_set, development_set=None):
...
+            if development_set is not None:
+                dev_accuracy = self.run_evaluation(development_set)
+                log_line = 'dev accuracy: {:3.2f}'.format(dev_accuracy)
+                log_line += '   epoch: {}'.format(epoch)
+                logger.info(log_line)
+
+                if not self.is_best(dev_accuracy):
+                    continue
+
+                # In case model perform better than previous model
```
Comment (jun-harashima) [inline msnc/model.py]: [nits][IMO] `when the new model outperforms the previous ones`

File: msnc/model.py [hunk 3]
```diff
@@ -96,14 +99,32 @@ def run_training(self, output_dir, training_set, development_set=None):
...
+                log_line = '[new best] dev accuracy: {:3.2f}'.format(dev_accuracy)  # NOQA
+                log_line += '   epoch: {}'.format(epoch)
 
-        log_line = 'best_accuracy: {:3.2f}'.format(self._best_accuracy)
-        log_line += '   best_epoch: {}'.format(self._best_epoch)
+                # Save epoch and performance informations
```
Comment (jun-harashima) [inline msnc/model.py]: `informations` -> `information`

File: msnc/model.py (hunk 1 above)
Comment (jun-harashima) [reply msnc/model.py]: looks great for me 👍

**Comment by jun-harashima** [comment]
LGTM w/ comments 🙇

File: msnc/model.py (hunk 3 above)
Comment (himkt) [reply msnc/model.py]: 📛

**Comment by himkt** [comment]
@jun-harashima Thanks for comments.
I revised and will merge right now. 🚀

**Comment by jun-harashima** [comment]
Deployed 🚀 Thank you for the PR 🎉
